### PR TITLE
Produce code coverage in CI

### PR DESCRIPTION
**Pull request by thomcc**
Sadly, this can't see a lot of test code, but is better than nothing.

### Pull Request checklist ###

- [x] **Quality**: This PR builds and tests run cleanly
  - `automation/all_tests.sh` runs to completion and produces no failures
  - Note: For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes a changelog entry in [CHANGES_UNRELEASED.md](../CHANGES_UNRELEASED.md) or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [x] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/main/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due dilligence applied in selecting them.

### CODE REVIEW

File: automation/emit_coverage_info.sh [hunk 1]
```diff
@@ -27,20 +28,16 @@ for manifest in $(cargo metadata --no-deps --format-version 1 | jq -r '.packages
...
     package=$(basename "$package")
     echo "## no-default-features test for package $package (manifest @ $manifest)"
-    cargo +nightly test --manifest-path "$manifest" --no-default-features
+    cargo +nightly test --manifest-path "$manifest" --all-targets --no-default-features --no-fail-fast
 done
 
-for bench_pkg in $(cargo metadata --format-version 1 --no-deps | jq -r '.packages[]  | select(.targets[] | select(.kind[] | contains("bench"))) | .name' | uniq); do
-    cargo +nightly test -p "$bench_pkg" --benches
-done
+# for bench_pkg in $(cargo metadata --format-version 1 --no-deps | jq -r '.packages[]  | select(.targets[] | select(.kind[] | contains("bench"))) | .name' | uniq); do
```
Comment (thomcc) [inline automation/emit_coverage_info.sh:34]: I think --all-targets covers this but we'll see.

Note that --all-targets does not run doctests :/ https://github.com/rust-lang/cargo/issues/6669

**Comment by thomcc** [comment]
Note: Coverage is low because of NSS CI build error :( Will dig later.

**Comment by thomcc** [comment]
Update: I believe the reasons for the failures here are that -Clink-dead-code (required for accurate coverage) is causing code that's usually DCEd to be linked in, and that code has linker errors.

**Comment by thomcc** [comment]
This is ready for review!

File: automation/emit_coverage_info.sh [hunk 2]
```diff
@@ -0,0 +1,57 @@
+#!/usr/bin/env bash
```
Comment (eoger) [inline automation/emit_coverage_info.sh:1]: This file is very well documented, thanks for taking the time to do it.

File: components/support/ffi/src/handle_map.rs [hunk 3]
```diff
@@ -1067,7 +1067,7 @@ mod test {
     use super::*;
 
     #[derive(PartialEq, Debug)]
-    struct Foobar(usize);
+    pub(super) struct Foobar(usize);
```
Comment (eoger) [inline components/support/ffi/src/handle_map.rs:1070]: til about `pub(super)`

File: automation/emit_coverage_info.sh [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+fi
+
+cargo +nightly clean
+# Note that incremental builds, as well as sccache seems to break coverage :(.
+# So, we disable those.
+export CARGO_INCREMENTAL=0
+unset RUSTC_WRAPPER
+# TODO: re-add `-Clink-dead-code` to RUSTFLAGS -- right now it causes a build
+# linker error in NSS, as apparently some part of NSS we don't use doesn't have
+# the build flags set up correctly. This makes our code coverage accuracy worse,
```
Comment (eoger) [inline automation/emit_coverage_info.sh:32]: > as apparently some part of NSS we don't use doesn't have the build flags set up correctly

Do we know which parts? I wouldn't mind opening an issue for that.

File: .circleci/config.yml [hunk 5]
```diff
@@ -251,6 +270,22 @@ jobs:
       - run: cargo clippy --version
       - run: bash automation/all_clippy_checks.sh
       - save-sccache-cache
+  Generate code coverage:
+    docker:
+      - image: circleci/rust:latest
+    resource_class: large
```
Comment (eoger) [inline .circleci/config.yml:276]: Note: this would increase our CI cost substantially (but it is way worth it)

Comment (thomcc) [reply .circleci/config.yml:276]: Looks like `large` instead of `xlarge` works...

Comment (thomcc) [reply .circleci/config.yml:276]: I've made normal tests use `large` too -- that's not really necessary to do in *this* PR, but it seems good to land the "reduce CI costs" patch at the same time as the "increase CI costs" patch.

File: automation/emit_coverage_info.sh (hunk 4 above)
Comment (thomcc) [reply automation/emit_coverage_info.sh:32]: This is the linker error:
```
  = note: /usr/bin/ld: /home/circleci/project/libs/desktop/linux-x86-64/nss/lib/libintel-gcm-wrap_c_lib.a(intel-gcm-wrap_c_lib.intel-gcm-wrap.o): in function `intel_aes_gcmInitCounter':
          intel-gcm-wrap.c:(.text.intel_aes_gcmInitCounter+0x3af): undefined reference to `PORT_SetError_stub'
          /usr/bin/ld: intel-gcm-wrap.c:(.text.intel_aes_gcmInitCounter+0x3d5): undefined reference to `PORT_SetError_stub'
          /usr/bin/ld: /home/circleci/project/libs/desktop/linux-x86-64/nss/lib/libintel-gcm-wrap_c_lib.a(intel-gcm-wrap_c_lib.intel-gcm-wrap.o): in function `intel_AES_GCM_CreateContext':
          intel-gcm-wrap.c:(.text.intel_AES_GCM_CreateContext+0x19): undefined reference to `PORT_ZAlloc_stub'
          /usr/bin/ld: intel-gcm-wrap.c:(.text.intel_AES_GCM_CreateContext+0xb6): undefined reference to `PORT_Free_stub'
          /usr/bin/ld: /home/circleci/project/libs/desktop/linux-x86-64/nss/lib/libintel-gcm-wrap_c_lib.a(intel-gcm-wrap_c_lib.intel-gcm-wrap.o): in function `intel_AES_GCM_DestroyContext':
          intel-gcm-wrap.c:(.text.intel_AES_GCM_DestroyContext+0x44): undefined reference to `PORT_Free_stub'
          /usr/bin/ld: /home/circleci/project/libs/desktop/linux-x86-64/nss/lib/libintel-gcm-wrap_c_lib.a(intel-gcm-wrap_c_lib.intel-gcm-wrap.o): in function `intel_AES_GCM_EncryptUpdate':
          intel-gcm-wrap.c:(.text.intel_AES_GCM_EncryptUpdate+0xed): undefined reference to `PORT_SetError_stub'
          /usr/bin/ld: intel-gcm-wrap.c:(.text.intel_AES_GCM_EncryptUpdate+0xfe): undefined reference to `PORT_SetError_stub'
          /usr/bin/ld: intel-gcm-wrap.c:(.text.intel_AES_GCM_EncryptUpdate+0x10f): undefined reference to `PORT_SetError_stub'
          /usr/bin/ld: /home/circleci/project/libs/desktop/linux-x86-64/nss/lib/libintel-gcm-wrap_c_lib.a(intel-gcm-wrap_c_lib.intel-gcm-wrap.o): in function `intel_AES_GCM_DecryptUpdate':
          intel-gcm-wrap.c:(.text.intel_AES_GCM_DecryptUpdate+0xa8): undefined reference to `NSS_SecureMemcmp_stub'
          /usr/bin/ld: intel-gcm-wrap.c:(.text.intel_AES_GCM_DecryptUpdate+0xd1): undefined reference to `PORT_SetError_stub'
          /usr/bin/ld: intel-gcm-wrap.c:(.text.intel_AES_GCM_DecryptUpdate+0xe2): undefined reference to `PORT_SetError_stub'
          /usr/bin/ld: intel-gcm-wrap.c:(.text.intel_AES_GCM_DecryptUpdate+0xf3): undefined reference to `PORT_SetError_stub'
          /usr/bin/ld: intel-gcm-wrap.c:(.text.intel_AES_GCM_DecryptUpdate+0x119): undefined reference to `PORT_SetError_stub'
          /usr/bin/ld: /home/circleci/project/libs/desktop/linux-x86-64/nss/lib/libintel-gcm-wrap_c_lib.a(intel-gcm-wrap_c_lib.intel-gcm-wrap.o): in function `intel_AES_GCM_EncryptAEAD':
          intel-gcm-wrap.c:(.text.intel_AES_GCM_EncryptAEAD+0x144): undefined reference to `PORT_SetError_stub'
          /usr/bin/ld: /home/circleci/project/libs/desktop/linux-x86-64/nss/lib/libintel-gcm-wrap_c_lib.a(intel-gcm-wrap_c_lib.intel-gcm-wrap.o):intel-gcm-wrap.c:(.text.intel_AES_GCM_EncryptAEAD+0x156): more undefined references to `PORT_SetError_stub' follow
          /usr/bin/ld: /home/circleci/project/libs/desktop/linux-x86-64/nss/lib/libintel-gcm-wrap_c_lib.a(intel-gcm-wrap_c_lib.intel-gcm-wrap.o): in function `intel_AES_GCM_DecryptAEAD':
          intel-gcm-wrap.c:(.text.intel_AES_GCM_DecryptAEAD+0xdb): undefined reference to `NSS_SecureMemcmp_stub'
          /usr/bin/ld: intel-gcm-wrap.c:(.text.intel_AES_GCM_DecryptAEAD+0x11a): undefined reference to `PORT_SetError_stub'
          /usr/bin/ld: intel-gcm-wrap.c:(.text.intel_AES_GCM_DecryptAEAD+0x12b): undefined reference to `PORT_SetError_stub'
          /usr/bin/ld: intel-gcm-wrap.c:(.text.intel_AES_GCM_DecryptAEAD+0x13c): undefined reference to `PORT_SetError_stub'
          /usr/bin/ld: intel-gcm-wrap.c:(.text.intel_AES_GCM_DecryptAEAD+0x162): undefined reference to `PORT_SetError_stub'
          collect2: error: ld returned 1 exit status
```

Never managed to trigger it on my linux laptop -- only in CI.

Comment (thomcc) [reply automation/emit_coverage_info.sh:32]: Filed https://github.com/mozilla/application-services/issues/3390

File: automation/emit_coverage_info.sh [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+    cargo +nightly test --manifest-path "$manifest" --no-default-features --no-fail-fast || true
+done
+
+cargo +nightly test --all --no-fail-fast || true
+
+cargo +nightly run -p sync-test || true
+cargo +nightly run -p protobuf-gen -- tools/protobuf_files.toml || true
+cargo +nightly run -p systest || true
+
+env RUSTFLAGS="--cfg __appsvc_ci_hack $RUSTFLAGS" cargo +nightly test --all --all-features --no-fail-fast || true
```
Comment (rfk) [inline automation/emit_coverage_info.sh:57]: What is `__appsvc_ci_hack`? (I mean I can kind of guess from the name, but I can't find an other uses of it in this PR).

Comment (thomcc) [reply automation/emit_coverage_info.sh:57]: That's because it isn't from this PR, but from https://github.com/mozilla/application-services/pull/3343.

**Pull request merged**